externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/ui/EngineTab.scala

```diff
@@ -24,25 +24,42 @@ import scala.util.control.NonFatal
 import org.apache.kyuubi.{Logging, Utils}
 import org.apache.kyuubi.config.KyuubiConf
 import org.apache.kyuubi.engine.spark.SparkSQLEngine
+import org.apache.kyuubi.engine.spark.events.EngineEventsStore
 import org.apache.kyuubi.service.ServiceState
 
 /**
  * Note that [[SparkUITab]] is private for Spark
  */
-case class EngineTab(engine: SparkSQLEngine)
-  extends SparkUITab(engine.spark.sparkContext.ui.orNull, "kyuubi") with Logging {
+case class EngineTab(
+    engine: Option[SparkSQLEngine],
+    sparkUI: Option[SparkUI],
+    store: EngineEventsStore,
+    kyuubiConf: KyuubiConf)
+  extends SparkUITab(sparkUI.orNull, "kyuubi") with Logging {
 
   override val name: String = "Kyuubi Query Engine"
-  val killEnabled = engine.getConf.get(KyuubiConf.ENGINE_UI_STOP_ENABLED)
+  val killEnabled = kyuubiConf.get(KyuubiConf.ENGINE_UI_STOP_ENABLED)
 
-  engine.spark.sparkContext.ui.foreach { ui =>
+  val startTime = engine.map(_.getStartTime).getOrElse {
+    sparkUI
+      .map(ui => ui.store.applicationInfo().attempts.head.startTime.getTime)
+      .getOrElse(0L)
+  }
+
+  def endTime(): Long = engine.map(_ => System.currentTimeMillis()).getOrElse {
+    sparkUI
+      .map(ui => ui.store.applicationInfo().attempts.head.endTime.getTime)
```

Comment:
why the start/end Time always use the first attempt？

Reply:
I see that the spark source code also uses `attempts.head` to get attempt.
like: https://github.com/apache/spark/blob/2a9416abd5c8d83901c09a08433bf91e649dc43f/core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala#L277